flash_f4.c

```diff
@@ -0,0 +1,123 @@
+#include <stdint.h>
+#include <stddef.h>
+#include "flash.h"
+
+/* Flash registers. Copied here to avoid dependencies on either libopencm3 or
+ * ChibiOS. */
+#define MMIO32(addr) (*(volatile uint32_t *)(addr))
+#define FLASH_MEM_INTERFACE_BASE 0x40023C00
+#define FLASH_ACR                MMIO32(FLASH_MEM_INTERFACE_BASE + 0x00)
+#define FLASH_KEYR               MMIO32(FLASH_MEM_INTERFACE_BASE + 0x04)
+#define FLASH_OPTKEYR            MMIO32(FLASH_MEM_INTERFACE_BASE + 0x08)
+#define FLASH_SR                 MMIO32(FLASH_MEM_INTERFACE_BASE + 0x0C)
+#define FLASH_CR                 MMIO32(FLASH_MEM_INTERFACE_BASE + 0x10)
+
+#define FLASH_KEY1               0x45670123
+#define FLASH_KEY2               0xCDEF89AB
+#define FLASH_CR_SNB_POS         3
+#define FLASH_CR_LOCK            (1 << 31)
+#define FLASH_CR_PSIZE           ((uint32_t)0x03 << 8)
+#define FLASH_CR_PG              (1 << 0)
+#define FLASH_CR_SNB             ((uint32_t)0x000000F8)
+#define FLASH_CR_SER             ((uint32_t)0x00000002)
+#define FLASH_CR_STRT            ((uint32_t)0x00010000)
+
+#define FLASH_SR_BSY             (1 << 16)
+
+uint8_t flash_addr_to_sector(void *p)
+{
+    uint32_t addr = (uint32_t)p;
+    uint8_t sector;
+    uint32_t offset = addr & 0xFFFFFF;
+    if (offset < 0x10000) {
+        sector = offset / 0x4000; // 16K sectors, 0x08000000 to 0x0800FFFF
+    } else if (offset < 0x20000) {
+        sector = 3 + offset / 0x10000; // 64K sector, 0x08010000 to 0x0801FFFF
+    } else {
+        sector = 4 + offset / 0x20000; // 128K sectors, 0x08010000 to 0x080FFFFF
+    }
+    if (addr >= 0x08100000) {
+        // Bank 2, same layout, starting at 0x08100000
+        sector += 12;
+    }
+    return sector;
+}
+
+void flash_lock(void)
+{
+    FLASH_CR |= FLASH_CR_LOCK;
+}
+
+void flash_unlock(void)
+{
+    // ensure cleared state
+    FLASH_CR |= FLASH_CR_LOCK;
```

Comment:
[question] This line looks the same as calling `flash_lock()`. Why lock before unlocking?

Reply:
It is done the same way in libopencm3: https://github.com/libopencm3/libopencm3/blob/master/lib/stm32/common/flash_common_f234.c#L56-L64

Basically the idea is that if you interrupt between the write to KEY1 & KEY2, then restart the sequence, the microcontroller will see KEY1 -> KEY1, abort the sequence, then see a write to KEY2 which is not a sequence as well. This is a weird shortcoming of the STM32 hardware state machine as far as I understand it.